data/store/mongo/mongo_datum.go

```diff
@@ -598,133 +579,219 @@ func (d *DatumRepository) CheckDataSetContainsTypeInRange(ctx context.Context, d
+	// quit early if range is 0
+	if status.FirstData.Equal(status.LastData) {
+		return nil, nil
+	}
+
+	// return error if ranges are inverted, as this can produce unexpected results
+	if status.FirstData.After(status.LastData) {
+		return nil, fmt.Errorf("FirstData (%s) after LastData (%s) for user %s", status.FirstData, status.LastData, userId)
 	}
 
 	// quit early if range is 0
-	if startTime.Equal(endTime) {
-		return nil
+	if status.LastUpdated.Equal(status.NextLastUpdated) {
+		return nil, nil
 	}
 
 	// return error if ranges are inverted, as this can produce unexpected results
-	if startTime.After(endTime) {
-		return fmt.Errorf("startTime (%s) after endTime (%s) for user %s", startTime, endTime, userId)
+	if status.LastUpdated.After(status.NextLastUpdated) {
+		return nil, fmt.Errorf("LastUpdated (%s) after NextLastUpdated (%s) for user %s", status.LastUpdated, status.NextLastUpdated, userId)
 	}
 
 	selector := bson.M{
 		"_active": true,
 		"_userId": userId,
 		"type":    typ,
 		"time": bson.M{
-			"$gt":  startTime,
-			"$lte": endTime,
+			"$gt":  status.FirstData,
+			"$lte": status.LastData,
+		},
+		"modifiedTime": bson.M{
+			"$gt":  status.LastUpdated,
+			"$lte": status.NextLastUpdated,
 		},
 	}
 
 	opts := options.Find()
 	opts.SetSort(bson.D{{Key: "time", Value: 1}})
+	opts.SetBatchSize(300)
 
 	cursor, err := d.Find(ctx, selector, opts)
 	if err != nil {
-		return fmt.Errorf("unable to get cgm data in date range for user: %w", err)
+		return nil, fmt.Errorf("unable to get %s data in date range for user: %w", typ, err)
 	}
 
-	if err = cursor.All(ctx, dataRecords); err != nil {
-		return fmt.Errorf("unable to decode data sets, %w", err)
+	return cursor, nil
+}
+
+func (d *DatumRepository) getTimeRange(ctx context.Context, userId string, typ string, status *types.UserLastUpdated) (err error) {
+	timestamp := time.Now().UTC()
+	futureCutoff := timestamp.AddDate(0, 0, 1)
+	pastCutoff := timestamp.AddDate(-2, 0, 0)
+
+	// get latest active record
+	selector := bson.M{
+		"_active": true,
+		"_userId": userId,
+		"type":    typ,
+		"time": bson.M{
+			"$gte": pastCutoff,
+			"$lte": futureCutoff,
+		},
 	}
 
-	return nil
-}
+	findOptions := options.Find()
+	findOptions.SetSort(bson.D{{Key: "time", Value: -1}})
+	findOptions.SetLimit(1)
 
-func (d *DatumRepository) GetLastUpdatedForUser(ctx context.Context, userId string, typ string) (*types.UserLastUpdated, error) {
-	var err error
 	var cursor *mongo.Cursor
-	var status = &types.UserLastUpdated{}
+	cursor, err = d.Find(ctx, selector, findOptions)
+	if err != nil {
+		return fmt.Errorf("unable to get last %s time: %w", typ, err)
+	}
+
 	var dataSet []*baseDatum.Base
+	if err = cursor.All(ctx, &dataSet); err != nil {
+		return fmt.Errorf("unable to decode last %s time: %w", typ, err)
+	}
 
-	if ctx == nil {
-		return nil, errors.New("context is missing")
+	// if we have a record
+	if len(dataSet) > 0 {
+		status.LastData = dataSet[0].Time.UTC()
```

Comment:
`getTimeRange` doesn't imply one of the arguments will get modified. It's better to return a new struct instead.

Reply:
renamed to populateTImeRange